src/redis-benchmark.c

```diff
@@ -605,17 +622,24 @@ static void writeHandler(aeEventLoop *el, int fd, void *privdata, int mask) {
     /* Initialize request when nothing was written. */
     if (c->written == 0) {
         /* Enforce upper bound to number of requests. */
-        int requests_issued = 0;
-        atomicGetIncr(config.requests_issued, requests_issued, config.pipeline);
-        if (requests_issued >= config.requests) {
-            return;
+        if (config.qps == 0) {
+            int requests_issued = 0;
+            atomicGetIncr(config.requests_issued, requests_issued, config.pipeline);
+            if (requests_issued >= config.requests) {
+                return;
+            }
         }
 
         /* Really initialize: randomize keys and set start time. */
         if (config.randomkeys) randomizeClientKey(c);
         if (config.cluster_mode && c->staglen > 0) setClusterKeyHashTag(c);
         atomicGet(config.slots_last_update, c->slots_last_update);
-        c->start = ustime();
+        if (config.qps) {
+            long long timestamp = popItem();
+            if (timestamp == -1) return;
```

Comment:
There might be a busy loop here, returning without performing any write will immediately call the write handler again.

Reply:
> I thought it would be processed in the next loop... So from my understanding, it may cause the calculated latency of other requests larger than actual value? 

It will be processed in the next loop, but as long as `popItem` returns `-1` you'll return and immediately call the write handler again. This will make the client thread work at 100% cpu even though there's nothing to do. I think `popItem` should block until there's more data to be sent. Does this make sense?

> I tried to fix it by deleting and recreating the writing event, but it seems that the latency increased instead...

I'm not sure you can know when to recreate the write event.

Reply:
>  I guess the reason is when `popItem` is blocked, `aeProcessEvents` in **ae.c** is also blocked and thus not able to process the coming received requests timely. ... I'm now trying another method to use `aeTimeEvent` to control the emit rate which will solve the problem of busy waiting and blocking if works well.

@liexusong This makes sense, if you have more than one client per thread then waiting on the queue might delay the other client's read/write handlers.

> I'm now trying another method to use aeTimeEvent to control the emit rate which will solve the problem of busy waiting and blocking if works well.

It'll be interesting to see if this works or has it own performance implications. I think that if we'll add a timer event per message it'll be problematic, if there's a good way to periodically add a timer event once we detect we need to slow down the QPS then this might work nicely. Alternatively you might want to try replacing your queue with an `anetPipe`, this way you can integrate it into the event loop without causing the delays you noticed.